### PR TITLE
std.fs.Dir: report truncation in readFile

### PR DESCRIPTION
**Pull request by notcancername**
This is a revised version of #18889 in response to https://github.com/ziglang/zig/pull/18889#issuecomment-1938026340 and #18880.

### CODE REVIEW

**Comment by andrewrk** [comment]
No good, the function must not make 2 syscalls.

**Comment by rootbeer** [comment]
Maybe just extend the comment to suggest work-arounds.  E.g., suggest passing a buffer with +1 length so you can detect if there is more data than you expected.  Or using a different set of API calls to read the file.  My expectation is that many callers are pretty confident in the worst-case size of the file content and don't need any overflow detection.  Do you have a use case in mind that benefits from the truncation detection?

**Comment by notcancername** [comment]
> No good, the function must not make 2 syscalls.

The best case is the same for the status quo and this PR at 3 syscalls, the worst case is `2 + buffer.len + 1` as opposed to `2 + buffer.len` syscalls. Rejecting this check based on one additional syscall that is made in the very narrow and improbable case where the entire buffer is used, which, based on how `readFile` is used, isn't intended by the caller, seems inconsistent. 

>  My expectation is that many callers are pretty confident in the worst-case size of the file content

I agree, but it's important that this assumption is actually verified because not doing so can lead to incorrect behavior.

>  Do you have a use case in mind that benefits from the truncation detection?

Based on how Zig uses `readFile`, truncation presents a problem. 

https://github.com/ziglang/zig/blob/0aff1f9ce51fd1b0df5cb5a6a06c5d6c744749fe/src/windows_sdk.zig#L753-L759

Here it's used to read a version which is then parsed. If the size of the version has actually grown beyond the intended size, it can still result in a valid version, which would silently be passed along. Ditto here:

https://github.com/ziglang/zig/blob/0aff1f9ce51fd1b0df5cb5a6a06c5d6c744749fe/lib/std/zig/system/darwin/macos.zig#L58-L69

The user who created #18880 seemed to have had a use case as well.

File: lib/std/fs/Dir.zig [hunk 1]
```diff
@@ -1740,16 +1740,23 @@ pub fn readLinkW(self: Dir, sub_path_w: []const u16, buffer: []u8) ![]u8 {
...
+/// of the file.
 pub fn readFile(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
     var file = try self.openFile(file_path, .{});
     defer file.close();
 
     const end_index = try file.readAll(buffer);
+
+    if (end_index == buffer.len) {
+        var unused: [1]u8 = undefined;
+        if (try file.read(&unused) > 0) {
```
Comment (daurnimator) [inline lib/std/fs/Dir.zig:1755]: This isn't the right thing to do for frame delimited formats (such as a file descriptor hooked up to an SCTP socket)

Comment (notcancername) [reply lib/std/fs/Dir.zig:1755]: You're correct, but I'm not aware of any way a `SOCK_SEQPACKET` or `SOCK_DGRAM` socket could possibly be opened with `openFile`. Feel free to correct me with a real-world proof-of-concept though.

Comment (daurnimator) [reply lib/std/fs/Dir.zig:1755]: > Feel free to correct me with a real-world proof-of-concept though.

As a trivial but not improbable example: `openFile` on `/proc/self/fd/N`

> using `readAll` would be invalid for non-stream sockets anyway

Howso? I'd expect `readAll` on an SOCK_SEQPACKET to return the entire current packet (but no more).

Comment (notcancername) [reply lib/std/fs/Dir.zig:1755]: > As a trivial but not improbable example: `openFile` on `/proc/self/fd/N`

It is not possible to open such a fd with `open`/`openat`, here is a demonstration of this fact:
```ts
const std = @import("std");

pub fn main() !void {
    var child = std.ChildProcess.init(&.{"ncat", "-vvvv", "--sctp", "-l", "-p", "6969"}, std.heap.c_allocator);
    defer _ = child.kill() catch {};

    try child.spawn();

    std.time.sleep(200 * std.time.ns_per_ms);

    const sock = try std.os.socket(std.os.AF.INET, std.os.SOCK.SEQPACKET, std.os.IPPROTO.SCTP);

    const sa = try std.net.Address.parseIp4("127.0.0.1", 6969);

    // set default address
    try std.os.connect(sock, @ptrCast(&sa.in), sa.getOsSockLen());


    var d = try std.fs.openDirAbsolute("/proc/self/fd/", .{});
    defer d.close();

    var b: [16]u8 = undefined;
    const sock_str = try std.fmt.bufPrint(&b, "{d}", .{sock});

    // must fail
    _ = d.openFile(sock_str, .{}) catch |err| std.debug.print("{}\n", .{err});
}
```

```
$ zig build-exe readfile_procself.zig -lc
$ ./readfile_procself 
Ncat: Version 7.94 ( https://nmap.org/ncat )
NCAT DEBUG: Initialized fdlist with 103 maxfds
Ncat: Listening on [::]:6969
NCAT DEBUG: Added fd 3 to list, nfds 1, maxfd 3
Ncat: Listening on 0.0.0.0:6969
NCAT DEBUG: Added fd 4 to list, nfds 2, maxfd 4
NCAT DEBUG: Added fd 0 to list, nfds 3, maxfd 4
NCAT DEBUG: Initialized fdlist with 100 maxfds
NCAT DEBUG: selecting, fdmax 4
NCAT DEBUG: unexpected errno: 6
select returned 1 fds ready
NCAT DEBUG: fd 4 is ready
NCAT DEBUG: Swapping fd[0] (3) with fd[2] (0)
NCAT DEBUG: Removed fd 3 from list, nfds 2, maxfd 4
NCAT DEBUG: Swapping fd[1] (4) with fd[1] (4)
NCAT DEBUG: Removed fd 4 from list, nfds 1, maxfd 0
Ncat: Connection from 127.0.0.1:60238.
NCAT DEBUG: Added fd 5 to list, nfds 2, maxfd 5
NCAT DEBUG: Added fd 5 to list, nfds 1, maxfd 5
NCAT DEBUG: selecting, fdmax 5
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/debug.zig:188:31: 0x106549c in dumpCurrentStackTrace (readfile_procself)
        writeCurrentStackTrace(stderr, debug_info, io.tty.detectConfig(io.getStdErr()), start_addr) catch |err| {
                              ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/os.zig:5831:40: 0x102b578 in unexpectedErrno (readfile_procself)
        std.debug.dumpCurrentStackTrace(null);
                                       ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/os.zig:1900:49: 0x106a4ec in openatZ (readfile_procself)
            else => |err| return unexpectedErrno(err),
                                                ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/fs/Dir.zig:843:26: 0x102d406 in openFileZ (readfile_procself)
        try posix.openatZ(self.fd, sub_path, os_flags, 0);
                         ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/fs/Dir.zig:775:26: 0x1028100 in openFile (readfile_procself)
    return self.openFileZ(&path_c, flags);
                         ^
/home/user/src/random/readfile_procself.zig:26:19: 0x10270a1 in main (readfile_procself)
    _ = d.openFile(sock_str, .{}) catch |err| std.debug.print("{}\n", .{err});
                  ^
/home/user/Downloads/zig-linux-x86_64-0.12.0-dev.2668+ddcea2cad/lib/std/start.zig:585:37: 0x1028706 in main (readfile_procself)
            const result = root.main() catch |err| {
                                    ^
../sysdeps/nptl/libc_start_call_main.h:58:16: 0x7fa6c3c9c6c9 in __libc_start_call_main (../sysdeps/x86/libc-start.c)
../csu/libc-start.c:360:3: 0x7fa6c3c9c784 in __libc_start_main_impl (../sysdeps/x86/libc-start.c)
???:?:?: 0x1025e90 in ??? (???)
???:?:?: 0x0 in ??? (???)
error.Unexpected
```

> Howso? I'd expect `readAll` on an SOCK_SEQPACKET to return the entire current packet (but no more).

`readAll` relies on the file being a stream, and reads until either the buffer is full or `read` returns 0:

https://github.com/ziglang/zig/blob/0c725a354a801c84100011db5eb53ae6c58086c9/lib/std/fs/File.zig#L1038-L1048

**Comment by rohlem** [comment]
Just as an idea, maybe instead of modifying the single `readFile` function we want to split it into several functions:
- the current behavior (`readFileFast`?)
- `readFileSmallerThanBuffer` which requires the buffer to be larger to check for truncation
- `readFileFullSize` which performs an additional syscall to check for a larger file
I think I personally agree that with `std.debug.runtime_safety == true` file truncation should be checked.
Otherwise it's probably up to `std` to decide the behavior of an unqualified `readFile` function, if it wants to provide one.

**Comment by notcancername** [comment]
`readFileFullSize` wouldn't be necessary with `readFileSmallerThanBuffer`, since that already checks for truncation, right?

I would welcome a `readFileTruncate` (status quo `readFile`) in addition to a `readFile` (which would ensure the entire file has been read, as in, this PR) to specify truncation, but I'd like to get some kind of consensus for such an addition first.

> EDIT: I guess this would still require an additional syscall for the happy case, right?

No, no, the additional syscall happens if and only if `buffer.len` is greater or equal to the file size. **Otherwise, the number of syscalls is identical.**

> I think I personally agree that with `std.debug.runtime_safety == true`, file truncation should be checked.

I don't think such an assertion would be useful, that would limit the reusability of the code. I favor your suggested approach.

**Comment by rohlem** [comment]
> `readFileFullSize` wouldn't be necessary with `readFileSmallerThanBuffer`, since that already checks for truncation, right?

My thinking was to stay compatible / support callers which only have/provide the exact buffer size, and can't allocate a larger buffer (might be managed externally).

>> EDIT: I guess this would still require an additional syscall for the happy case, right?

> No, no, the additional syscall happens if and only if buffer.len is greater or equal to the file size. Otherwise, the number of syscalls is identical.

Right, my idea for `readFileSmallerThanBuffer` was that the user submits a buffer larger than they expect the file to be.
So if the file is, as expected, smaller than the buffer, it's not clear whether the OS arbitrarily stopped the read operation at the expected file size, or because the file ended there.
In other words, we have worst-case up to `larger_buffer.len` syscalls instead of `smaller_buffer.len + 1` syscalls (which `readFileFullSize` would perform) - only the "unhappy case" of a too-large file can be detected faster if the OS fills more of the buffer than the expected file size in one syscall.

**Comment by notcancername** [comment]
That's a good idea, but I don't know if it's general enough to be in the standard library. `readFileTruncate`/status quo could do the trick just as well with an additional `if (returned.len > expected)` check, right?

**Comment by notcancername** [comment]
In response to your feedback, @rohlem, I've re-added the original implementation as `readFileTruncate` with a clarified doc comment to allow for truncation and reading with a matching buffer size without the truncation check.

**Comment by iacore** [comment]
> No good, the function must not make 2 syscalls.

It **must** if it wants to handle short reads.

**Comment by notcancername** [comment]
> On Linux, you can actually use `readv` to read to two buffers at once.

Thanks for the suggestion! I've updated the PR to use `readv`, making the best case of 3 syscalls a bit more likely :).

File: lib/std/fs/Dir.zig [hunk 2]
```diff
@@ -1740,17 +1740,39 @@ pub fn readLinkW(self: Dir, sub_path_w: []const u16, buffer: []u8) ![]u8 {
...
+
+    var unused: [1]u8 = undefined;
+    var iovecs = [_]std.os.iovec{
+        .{ .iov_base = buffer.ptr, .iov_len = buffer.len },
+        .{ .iov_base = &unused, .iov_len = unused.len },
+    };
+
+    const nb_read = try file.readvAll(&iovecs);
+
+    if (nb_read > buffer.len) {
```
[the remark this answers is not in the document]
Comment (notcancername) [reply lib/std/fs/Dir.zig:1771]: > When used on a pipe, this causes data loss

Please demonstrate a case where the file is immediately opened, then closed, and data loss is important.

> im not convinced the fix is an improvement over the bug

Then please suggest a better fix.

File: lib/std/fs/Dir.zig [hunk 3]
```diff
@@ -1740,17 +1740,39 @@ pub fn readLinkW(self: Dir, sub_path_w: []const u16, buffer: []u8) ![]u8 {
...
+/// Read the entirety of the file at `file_path` into `buffer`.
+/// The returned slice has the same pointer as `buffer`. If the entirety of the file does not fit
+/// into `buffer`, `error.Truncated` is returned, and `buffer` contains the first `buffer.len` bytes
+/// of the file.
+pub fn readFile(self: std.fs.Dir, file_path: []const u8, buffer: []u8) ![]u8 {
+    var file = try self.openFile(file_path, .{});
+    defer file.close();
+
+    var unused: [1]u8 = undefined;
+    var iovecs = [_]std.os.iovec{
```
[the remark this answers is not in the document]
Comment (notcancername) [reply lib/std/fs/Dir.zig:1764]: > readv tends to be slower than read

Demonstrate that with a benchmark, please.

File: lib/std/fs/Dir.zig (hunk 2 above)
[the remark this answers is not in the document]
Comment (notcancername) [reply lib/std/fs/Dir.zig:1771]: I appreciate your commitment to ensuring the stdlib works well even in edge cases, but it would have been nice to test this yourself first. In my tests with a FIFO and the `readv`, and truncating variants, the writing process always gets `SIGPIPE`'d as soon as the reading process closes the file, data loss is not a concern. In the common/intended case, where the FIFO contains less than `buffer.len` bytes, the implementations are indistinguishable to the reader:

<details> <summary>Implementations</summary>

`read_file_truncate.zig`
```zig
const std = @import("std");

const Dir = std.fs.Dir;

pub fn readFileTruncate(self: Dir, file_path: []const u8, buffer: []u8) ![]u8 {
    var file = try self.openFile(file_path, .{});
    defer file.close();

    const nb_read = try file.readAll(buffer);

    return buffer[0..nb_read];
}

pub fn main() !void {
    var buffer: [6969]u8 = undefined;
    _ = try readFileTruncate(std.fs.cwd(), "myfifo", &buffer);
}
```

`read_file_readv.zig`
```zig
const std = @import("std");

const Dir = std.fs.Dir;

pub fn readFileReadv(self: std.fs.Dir, file_path: []const u8, buffer: []u8) ![]u8 {
    var file = try self.openFile(file_path, .{});
    defer file.close();

    var unused: [1]u8 = undefined;
    var iovecs = [_]std.os.iovec{
        .{ .iov_base = buffer.ptr, .iov_len = buffer.len },
        .{ .iov_base = &unused, .iov_len = unused.len },
    };

    const nb_read = try file.readvAll(&iovecs);

    if (nb_read > buffer.len) {
        return error.Truncated;
    }

    return buffer[0..nb_read];
}

pub fn main() !void {
    var buffer: [6969]u8 = undefined;
    _ = try readFileReadv(std.fs.cwd(), "myfifo", &buffer);
}
```

</details>

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "00000000: 0000 0000 0000 0000 00"..., 6969) = 68
read(3, "", 6901)                       = 0
close(3)                                = 0
+++ exited with 0 +++
```
```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="00000000: 0000 0000 0000 0000 00"..., iov_len=6969}, {iov_base="", iov_len=1}], 2) = 68
readv(3, [{iov_base="", iov_len=6901}, {iov_base="", iov_len=1}], 2) = 0
close(3)                                = 0
+++ exited with 0 +++
```

In the intended but unlikely case where the file has exactly `buffer.len` bytes, the `readv` variant uses 4 syscalls: 

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., 6969) = 6969
close(3)                                = 0
+++ exited with 0 +++
```

```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"..., iov_len=6969}, {iov_base="", iov_len=1}], 2) = 6969
readv(3, [{iov_base="", iov_len=1}], 1) = 0
close(3)                                = 0
+++ exited with 0 +++
```
In the unintended case where the file has more than `buffer.len` bytes, the truncating variant behaves incorrectly in 3 syscalls:

```
$ strace -e trace=openat,close,read,readv ./read_file_truncate
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
read(3, "00000000: 0000 0000 0000 0000 00"..., 6969) = 6969
close(3)                                = 0
+++ exited with 0 +++
```

The `readv` variant behaves correctly, reading one more byte than the buffer in 3 syscalls:

```
$ strace -e trace=openat,close,read,readv ./read_file_readv
openat(AT_FDCWD, "myfifo", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 3
readv(3, [{iov_base="00000000: 0000 0000 0000 0000 00"..., iov_len=6969}, {iov_base="0", iov_len=1}], 2) = 6970
close(3)                                = 0
error: Truncated
(..., stacktrace follows)
```

File: lib/std/fs/Dir.zig (hunk 3 above)
[the remark this answers is not in the document]
Comment (iacore) [reply lib/std/fs/Dir.zig:1764]: musl libc actually uses `readv` only for the `FILE*` API.

**Comment by andrewrk** [comment]
I'm sorry, I didn't review this in time, and now it has bitrotted. Furthermore, so many pull requests have stacked up that I can't keep up and I am therefore declaring Pull Request Bankruptcy and closing old PRs that now have conflicts with master branch.

If you want to reroll, you are by all means welcome to revisit this changeset with respect to the current state of master branch, and there's a decent chance your patch will be reviewed the second time around.

Either way, I'm closing this now, otherwise the PR queue will continue to grow indefinitely.
